quinn/src/connection.rs

```diff
@@ -70,29 +70,55 @@ impl Connecting {
 
     /// Convert into a 0-RTT or 0.5-RTT connection at the cost of weakened security
     ///
-    /// Opens up the connection for use before the handshake finishes, allowing the API user to
-    /// send data with 0-RTT encryption if the necessary key material is available. This is useful
-    /// for reducing start-up latency by beginning transmission of application data without waiting
-    /// for the handshake's cryptographic security guarantees to be established.
+    /// Returns `Ok` immediately if the local endpoint is able to attempt 0/0.5-RTT conversion. If
+    /// so, the returned [`Connection`] can be used to open streams and write data to them without
+    /// waiting for the rest of the handshake to complete, at the cost of weakened cryptographic
+    /// security guarantees. The returned [`ZeroRttAccepted`] future resolves when the handshake
+    /// does complete, at which point subsequently opened streams and written data will have full
+    /// cryptographic protection.
     ///
-    /// When the `ZeroRttAccepted` future completes, the connection has been fully established.
+    /// Outgoing
+    /// ---
     ///
-    /// # Security
+    /// For outgoing connections, the initial attempt to convert to 0-RTT will proceed if the
+    /// [`crypto::ClientConfig`][crate::crypto::ClientConfig] attempts to resume a previous TLS
+    /// session. However, **the remote endpoint may not actually _accept_ the 0-RTT data**--yet
+    /// still accept the connection attempt in general. This may occur, for example, if the remote
```

Comment:
While I agree we should be technically correct, I also think it's pedagogically important here to ground it in a concrete and common-to-encounter example (server restart). If I had slightly less background knowledge here I might not know what a server "losing resumption state" means or whether it was something common or esoteric. Also, if my understanding is correct, the condition described in this sentence could occur in cases other than lost resumption state--even, hypothetically, a server for some reason being programmed to reject early data based on a random coin toss.

I'm trying to think of ways to rephrase this to make it feel more general. Perhaps: "One common reason this may occur is if the remote endpoint has restarted and thus lost its resumption state since the local endpoint last connected to it." I feel like that's a lot more clunky, grammatically convoluted, and lengthy, though. What do you think?

Reply:
Maybe something like:

"However, the server may at its discretion decide to reject 0-RTT data while otherwise accepting the connection. The ability to accept 0-RTT data depends on resumption state being stored in the server, which servers will limit due to any number of reasons. Some servers only store resumption state in memory, such that a restart of the server will reset all resumption state."

Reply:
I changed it to something based on that.